domain/src/main/java/org/oppia/domain/audio/AudioPlayerController.kt

```diff
@@ -0,0 +1,174 @@
+    }
+  }
+
+  private fun prepareDataSource(url: String) {
+    mediaPlayer.setDataSource(context, Uri.parse(url))
+    mediaPlayer.prepareAsync()
+    playState.postValue(AsyncResult.pending())
+  }
+
+  /** Call function to play audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in paused state */
+  fun play() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (!mediaPlayer.isPlaying) {
+      mediaPlayer.start()
+      scheduleNextSeekBarUpdate()
+    }
+  }
+
+  /** Call function to pause audio.
+   * Must call [initializeMediaPlayer] and wait for prepared state first.
+   * MediaPlayer should be in playing state */
+  fun pause() {
+    check(prepared) { "Media Player not in a prepared state" }
+    if (mediaPlayer.isPlaying) {
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PAUSED, mediaPlayer.currentPosition, mediaPlayer.duration)))
+      mediaPlayer.pause()
+      stopUpdatingSeekBar()
+    }
+  }
+
+  private fun scheduleNextSeekBarUpdate() {
+    if (observerActive && prepared) {
+      seekBarLock.withLock {
+        nextUpdateJob = CoroutineScope(backgroundDispatcher).launch {
+          delay(SEEKBAR_UPDATE_FREQUENCY)
+          updateSeekBar()
+          scheduleNextSeekBarUpdate()
+        }
+      }
+    }
+  }
+
+  private fun updateSeekBar() {
+    if (mediaPlayer.isPlaying) {
+      playState.postValue(AsyncResult.success(PlayProgress(PlayStatus.PLAYING, mediaPlayer.currentPosition, mediaPlayer.duration)))
+    }
+  }
+
+  private fun stopUpdatingSeekBar() {
+    seekBarLock.withLock {
+      nextUpdateJob?.cancel()
+      nextUpdateJob = null
+    }
+  }
+
+  /* Call function to release media player and stop seek bar updates
+  *  MediaPlayer must be active when calling this function */
+  fun releaseMediaPlayer() {
+    if (!mediaPlayerActive) throw MediaPlayerNotInitException()
+    mediaPlayerActive = false
+    prepared = false
+    mediaPlayer.release()
+    stopUpdatingSeekBar()
+  }
+
+  /** Call function to check playing state
+  *  Must call [initializeMediaPlayer] and wait for prepared state first. */
+  fun isPlaying() : Boolean {
+    check(prepared) { "Media Player not in a prepared state" }
+    return mediaPlayer.isPlaying
+  }
+
+  /** Call function to change progress of MediaPlayer
+   *  Must call [initializeMediaPlayer] and wait for prepared state first. */
+  fun seekTo(position: Int)  {
+    check(prepared) { "Media Player not in a prepared state" }
+    mediaPlayer.seekTo(position)
+  }
+
+  /* Observe to get updates on MediaPlayer current Progress
+  *  Prepared: Set duration of seek bar and attach listeners to UI
+  *  Playing: Update seek bar with current position
+  *  Paused: Change to pause icon
+  *  Completed: Reset play button and seek bar to original state */
+  fun getPlayState() : LiveData<AsyncResult<PlayProgress>> = playState
```

Comment:
Perhaps: ``getPlayProgress()``? ``PlayProgress`` is actually what's being returned here.

Ditto for the ``playState`` field name.

Reply:
Updated